### DISCUSSION TITLE
The reload speed

### DISCUSSION THREAD

**Discussion by v479038280** [zauberzeug/nicegui#1193] [General]
why the reload speed is not that fast than streamlit? I almost don't need to wait in in streamlit.

Another ,the default apperance of the widget is simple but not that beautiful.

**Comment by rodja** [comment]
> why the reload speed is not that fast than streamlit? I almost don't need to wait in in streamlit.

Interesting. I've not done a proper performance comparison yet. Could you share your Streamlit and NiceGUI code and how you measured it?

> Another ,the default apperance of the widget is simple but not that beautiful.

We wanted the UI elements to be as frictionless as possible. Hence we stick to the Quasar defaults as good as possible. That provides the least surprises if you want to adapt. But there are a few projects which build on top of NiceGUI to offer more styled components. Like https://github.com/ofenbach/CatDesign for example.

**Reply by v479038280** [reply]
actually ,I almost don't need to wait in streamlit, I save the code ,and then switch to the brower, the result is there.But when using niceGui ,I need to wait seconds. I just run the demo from the Document. you could try  almost any code in the tutorials ,the results are same.

**Reply by v479038280** [reply]
The element in QUASAR  is wonderful, and the use with tailwindcss make the UI more powerfull than streamlit .It's Just the default apperance looks not that good.   I should spend some time on tailwindcss and quasar.

**Comment by rodja** [comment]
> I just run the demo from the Document. you could try almost any code in the tutorials ,the results are same.

For me it's quite fast. Could you try a bare FastAPI app with uvicorn and reload=True? This is what NiceGUI is using under the hood. That would help us to determine if it's a NiceGUI issue (specific to your system) or more Uvicorn related...

**Reply by v479038280** [reply]
here is a simple hello world  with nicegui.

![ezgif com-video-to-gif](https://github.com/zauberzeug/nicegui/assets/25096045/f6872f4b-217b-4ec1-a2aa-44b07aaef2cc)

Here is  uvicorn ,though I pressed the refresh, I almost can't sense the waiting time (streamlit is like this)
![ezgif-5-e7cecd0a60](https://github.com/zauberzeug/nicegui/assets/25096045/22625a3b-dc5e-4cce-b546-160c836bf66b)

**Comment by rodja** [comment]
Wow, that is really slow and unexpected. This is how it looks on my side:

![Aug-04-2023 07-20-56](https://github.com/zauberzeug/nicegui/assets/131391/5918e369-acb3-4bda-a086-3d7c7ecb4906)

Is there some logging in the console? Can you share the debug log from the browser?

**Reply by v479038280** [reply]
here it is
![ezgif com-video-to-gif](https://github.com/zauberzeug/nicegui/assets/25096045/486bcf8c-1de3-40ff-9ab1-7b4bfdb591aa)

**Reply by falkoschindler** [reply]
~~The link is broken.~~

**Reply by v479038280** [reply]
It is OK now, but not that clear

**Reply by v479038280** [reply]
The reloading takes some time ,There is nothing wrong in the log output ,only a warn `WARNING:  WatchFiles detected changes in 'hello_world.py'. Reloading...`

**Reply by rodja** [reply]
Can you open the debug view in the browser? JS console and network traffic would be interesting.

**Reply by v479038280** [reply]
Is this one usefull for analyse
![GIF 2023-08-04 18-58-06](https://github.com/zauberzeug/nicegui/assets/25096045/5d07ba6a-6a03-4e9e-8681-e4c388171bf0)

**Comment by rodja** [comment]
Yes that may give us some leads. First I'm wondering why you get a TypeError in `all.js`:

![Screen Shot 2023-08-05 at 07 29 15](https://github.com/zauberzeug/nicegui/assets/131391/a34484a2-d634-46f1-b5b3-e34228bb8990)

We do not serve an `all.js` with NiceGUI. Where does it come from.

Second, you get a WebSocket connection error which is unexpected:

![Screen Shot 2023-08-05 at 07 32 56](https://github.com/zauberzeug/nicegui/assets/131391/437d5c48-a022-4339-ad4e-5dec14217539)

And the message "reloading because handshake failed" also indicates something wrong with the socket connection.

But we should first investigate the `all.js`. Maybe the socket error is caused by this.

**Reply by v479038280** [reply]
The all.js is from chrome-extension, look at the picture below. 
![微信图片编辑_20230805143323](https://github.com/zauberzeug/nicegui/assets/25096045/a64db7dc-74ee-4fb8-a2fd-0a2b30fa76ed)

Here is another demo running in firefox. No all.js
![GIF 2023-08-05 14-37-36](https://github.com/zauberzeug/nicegui/assets/25096045/8437170d-fea1-4efa-a170-4957a9f948d2)

 "reloading because handshake failed" 
![微信截图_20230805144537](https://github.com/zauberzeug/nicegui/assets/25096045/b285e2a9-da30-46ea-a5be-cd12e6403a0a)

**Comment by rodja** [comment]
It's very strange to have a failing handshake. Can you try running the docker image with your code?

```bash
docker run --rm -it -p 8080:8080 -v $(pwd):/app/ zauberzeug/nicegui:1.3.8
```

With `-v` your local folder is mounted as `/app`. If there is a main.py you should then be able to modify it to trigger a reload.

**Reply by v479038280** [reply]
Is the speed normal?
![GIF 2023-08-05 14-37-36](https://github.com/zauberzeug/nicegui/assets/25096045/488d33b5-3d58-45b4-9077-8dfc310dc41a)

**Reply by rodja** [reply]
Yes, thats much faster than what you showed before. There is also no failing handshake. So the problem is with your installation. Maybe remove NiceGUI (and dependencies) and reinstall?

**Reply by v479038280** [reply]
I'm now using nicegui 1.3.7, I update it to 1.3.8.  There is still a handshake failed ,but the waiting time is much shorter than before.The experience was much better than before

![微信截图_20230808165359](https://github.com/zauberzeug/nicegui/assets/25096045/8fea33d5-5c30-44bb-8877-b23254de3168)

**Comment by aersam** [comment]
Hi there
I'm also having issues with "reloading because handshake failed" 

Is it possible there are some socket.io timeouts which are too low? or can we configure those? 

I'm behind an azure website, so the installtion might be the problem, but I think it's unlikely . I have the issue on windows and on my linux azure website (which runs behind an IIS, tough).

**Reply by v479038280** [reply]
I run Nicegui on several computer ,it seems all with such problem

**Reply by rodja** [reply]
The "handshake failed" error should not appear on every start. Too short timeouts are unlikely. The message is printed here:

https://github.com/zauberzeug/nicegui/blob/219da1118c30d0611ec665cb7cf8b185b36833de/nicegui/templates/index.html#L241-L245

As you can see it only appears if the handshake event comes back falsy. Here is the backend code

https://github.com/zauberzeug/nicegui/blob/219da1118c30d0611ec665cb7cf8b185b36833de/nicegui/air.py#L64-L73

It seems like the client id is not found.

Maybe it's a Windows thing? Please try 

```bash
docker run --rm -it -p 8080:8080 -v $(pwd):/app/ zauberzeug/nicegui:1.3.13
```

To test if the same error appears with docker.

**Reply by aersam** [reply]
The thing is that it occurs on my linux azure website as well as on my windows machine. It does not happen on every start, but sometimes twice or even three or four times. 
I cannot see failed network requests, so it's really a bit strange.

Do you think doing a basic retry of the handshake this could help? As I really don't understand the reason I'm unsure

```javascript
window.socket.emit("handshake", (ok) => { 
   if (!ok) { 
     setTimeout(()=>
           window.socket.emit("handshake", (ok) => { 
                 if (!ok) { 
                         console.log('reloading because handshake failed') 
                         window.location.reload(); 
                  } 
        })
      , 200)
   } 
```

**Reply by rodja** [reply]
> It does not happen on every start, but sometimes twice or even three or four times. I cannot see failed network requests, so it's really a bit strange.

Very interesting clue. Thanks for mentioning. Has your azure website multiple instances running? If a load balancer is routing traffic to multiple servers the websocket connection may not be send to the same instance which created the html page. That would explain the  irregularity. Of course, with a retry you could fire connection requests until you get lucky and the request is send to the right server. A better solution would be do activate some kind of session affinity (eg. that the load balancer routes traffic from one client always to the same server instance).

Do you see the same irregularity on windows with a local started NiceGUI app?

**Reply by aersam** [reply]
Oh of course! We have gunicorn with multiple workers in place. We'll remove that

**Reply by aersam** [reply]
Removing gunicorn makes it work, which makes total sense. However I think supporting something like gunicorn would be pretty simple, I'll open another discussion for that :)

**Reply by rodja** [reply]
Thanks @aersam for starting discussion about supporting "multiple workers" in https://github.com/zauberzeug/nicegui/discussions/1539. Does that also explain the strange behaviour on Windows or only the Azure setup. What about you @v479038280? Does that explanation work for you, too?

**Reply by v479038280** [reply]
Actually I don't know how to remove gunicorn..

**Reply by rodja** [reply]
Can you share a minimal example of the code you are using which reproduces the problem?

**Reply by aersam** [reply]
> Actually I don't know how to remove gunicorn..

Setting number of workers to one works as well 🙂

**Reply by rodja** [reply]
Wonderful. I'll close this discussion as solved. How to support multiple workers will be tackled in https://github.com/zauberzeug/nicegui/discussions/1539.
